gms/gossiper.cc

```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
     });
 }
 
-future<> gossiper::do_shadow_round() {
-    return seastar::async([this, g = this->shared_from_this()] {
-        build_seeds_list();
-        _in_shadow_round = true;
-        auto t = clk::now();
-
-        // When peer node receives a syn message, it will send back a ack message.
-        // So, we need to register gossip message handlers before sending syn message.
+future<> gossiper::do_shadow_round(std::unordered_set<gms::inet_address> nodes) {
+    return seastar::async([this, g = this->shared_from_this(), nodes = std::move(nodes)] () mutable {
         get_gossiper().invoke_on_all([] (gossiper& g) {
             return g.init_messaging_service_handler();
         }).get();
 
-        while (this->_in_shadow_round) {
-            // send a completely empty syn
-            for (inet_address seed : _seeds) {
-                utils::chunked_vector<gossip_digest> digests;
-                gossip_digest_syn message(get_cluster_name(), get_partitioner_name(), digests);
-                auto id = get_msg_addr(seed);
-                logger.trace("Sending a GossipDigestSyn (ShadowRound) to {} ...", id);
-                // Do it in the background.
-                (void)ms().send_gossip_digest_syn(id, std::move(message)).handle_exception([id] (auto ep) {
-                    logger.trace("Fail to send GossipDigestSyn (ShadowRound) to {}: {}", id, ep);
-                });
+        nodes.erase(get_broadcast_address());
+
+        gossip_get_endpoint_states_request request{{
+            gms::application_state::STATUS,
+            gms::application_state::HOST_ID,
+            gms::application_state::TOKENS,
+            gms::application_state::SUPPORTED_FEATURES}};
+        logger.info("The do_shadow_round started, nodes={}", nodes);
```

Comment:
Q: how do we learn about supported features of nodes that are down? without that, we cannot join a cluster that has a node down.

Reply:
We learn them from the system  peer  table. So the features check is a combination of gossip live data + data from system tables.